### ISSUE TITLE
Monitor Health for a Reference Patroni installation using a Canary Application

### ISSUE THREAD

**Issue by wmhutchison** [BCDevOps/developer-experience#2966]
**Describe the issue**

This EPIC will track efforts towards monitoring the health of a highly available database deployment via the use of a suitable Canary application which will fail or return an appropriate message/code when it does not work as expected. The intent is to ensure that official images of the database software involved (for this EPIC, that will be Patroni) are monitored to ensure they work as expected during regular maintenance activities like Openshift upgrades, firmware upgrades for clusters using physical nodes, or just adhoc node drains. Unplanned events such as node failures for whatever reason is also covered off, but the value-add for this monitor is to ensure the Patroni software for current/future upgrades is fully able to maintain HA during normal maintenance activities.

**General Design Considerations**
- Canary Application should be written to be re-used for monitoring other database technologies.
- Canary Application should check a number of key points for a database, and return a FAIL if any of them are met.
- Canary Application should not be monitored with a "middle man" technology such as Sysdig as it introduces additional complexity and reliance upon Sysdig data and its out-sourced SaaS to be functioning.
- Canary Application monitoring would be best served with a localized technology such as DXC's Nagios monitoring, or the target cluster's AlertManager system. For MVP, Nagios monitoring will be what is pursued.

**Specific Requirements for Canary Monitoring of Patroni**
Canary App should return a FAIL if any of the following checks fail.
- check to make sure one can connect to the DB. Addresses checks involving DB cluster stability in whatever form that takes.
- check to make sure one can write an expected data point to the DB. Addresses issues involving storage being functional.
- check to make sure one can then read back the expected data point from the DB. Addresses potential corruption issues if we write but when reading do not get back the same content.

**MVP Top-Level Requirements**
- Monitoring of the Canary Application will alert when there are issues writing/reading from a database or otherwise connecting to the database. If all are successful, then no alerts will be sent.
- Canary Application will be initially setup using DXC's Nagios monitoring. Additional monitoring via uptime.com or similar methods will be considered later as backup.
- Canary Application will be monitoring known stable Patroni clusters. At present this will be just the single supported Patroni image available in Artifactory or the bcgov namespace.
- Canary Application will be stand-alone to its own container instead of being a single container trying to execute multiple checks of different services. Reasons for this include:
  - Proper monitoring means the canary app itself is not what does the alerting, thus the output of calling the canary app needs to be simple and not hard to parse out. A single URL that returns a specific code/message is easily setup on any monitoring solution we desire.
  - Putting too many checks into a single Canary app runs the risk of something going sideways for one check and affecting the results of the checks not yet performed.
  - cost of multiple Canary app pods is trivial compared to the resource cost involved with setting up one or more database clusters. Also improves visibility to look at a list of appropriately named canary pods to be reminded of what is being monitored.

While it is possible to expose and obtain more detailed information from the Patroni API, trying to go this route will result in excessive complexity in terms of trying to create a more full-fledged automated troubleshooting tool. Perhaps long-term such could be explored, but for now, we need only worry about ensuring suitable people are notified when something is "wrong" with this specific Canary App.

**Definition of Done**
- [x] Finalize MVP with Cailey via meetings and other discussions on Rocket Chat.
- [ ] Setup Proof of Concept of both a reference Patroni cluster and working Canary App for testing the health of said cluster, using an existing LAB cluster.
- [ ] Setup Automation of Deployment via CCM or other means to ensure both Canary app and Patroni clusters are suitably maintained.

**Comment by wmhutchison** [comment]
PoC initial Python script now created which will generate an exemption when either any of the DB operations fail, or if the requested DB query returns a count of 0, meaning that the data previously written into the DB isn't available. This can now be re-factored to start being ported into a container-native app that'll read senstive DB login from secrets, and output a suitable value for Nagios and/or third party monitoring tools.

```
import psycopg2

try:
  conn = psycopg2.connect(dbname="registry",
                        user="app_api_yoh51nj7",
                        host="10.98.55.92",
                        password="so78l0e4crpj68vxvowrynvc02t5qg0n",
                        port="5432",
                        connect_timeout="2")
  cursor = conn.cursor()
  cursor.execute("CREATE TABLE IF NOT EXISTS dummy_table(name varchar(20), value int);")
  cursor.execute("INSERT INTO dummy_table values('platypus',1);")
  conn.commit()
  cursor.execute("SELECT COUNT(*) FROM dummy_table WHERE name='platypus';")
  result = cursor.fetchone()
  canary_rows = result[0]
  if canary_rows == 0:
    raise Exception("No valid DB entries found in Canary database")
  cursor.execute("DELETE FROM dummy_table WHERE name='platypus';")
  conn.commit()
except:
  print("Canary could not talk to the Patroni cluster")
else:
  print(canary_rows)
  conn.close()
```

**Comment by rstens** [comment]
In the framework that I am proposing, this would be very easy to implement with almost no code (except for the SQL). So the need for a canary app would not be there. Happy to demonstrate.
